core/commonMain/src/implementations/immutableSet/TrieNode.kt

```diff
@@ -277,6 +304,94 @@ internal class TrieNode<E>(
         return element == buffer[cellIndex]
     }
 
+    fun addAll(otherNode: TrieNode<E>, shift: Int, intersectionSizeRef: DeltaCounter): TrieNode<E> {
+        if (this === otherNode) {
+            intersectionSizeRef += calculateSize()
+            return this
+        }
+        if (shift > MAX_SHIFT) {
+            return collisionAddAll(otherNode, intersectionSizeRef)
+        }
+        // union mask contains all the bits from input masks
+        val newBitMap = bitmap or otherNode.bitmap
+        // first allocate the node and then fill it in
+        // we are doing a union, so all the array elements are guaranteed to exist
+        val newNode = TrieNode<E>(newBitMap, arrayOfNulls<Any?>(newBitMap.countOneBits()))
+        // for each bit set in the resulting mask,
+        // either left, right or both masks contain the same bit
+        // Note: we shouldn't overrun MAX_SHIFT because both sides are correct TrieNodes, right?
+        newBitMap.forEachOneBit { positionMask ->
+            val newNodeIndex = newNode.indexOfCellAt(positionMask)
+            val thisIndex = indexOfCellAt(positionMask)
+            val otherNodeIndex = otherNode.indexOfCellAt(positionMask)
+            newNode.buffer[newNodeIndex] = when {
+                // no element on left -> pick right
+                hasNoCellAt(positionMask) -> otherNode.buffer[otherNodeIndex]
+                // no element on right -> pick left
+                otherNode.hasNoCellAt(positionMask) -> buffer[thisIndex]
+                // both nodes contain something at the masked bit
+                else -> {
+                    val thisCell = buffer[thisIndex]
+                    val otherNodeCell = otherNode.buffer[otherNodeIndex]
+                    val thisIsNode = thisCell is TrieNode<*>
+                    val otherIsNode = otherNodeCell is TrieNode<*>
```

Comment:
Does extracting these `is` checks into variables brings benefits here? 
We have to repeat the cast again in the when branches below (I suppose due to this smart-cast not being supported by Kotlin DFA).

Reply:
Well, this way we do it two times (check here and cast later), but otherwise we (potentially) do it up to four times (if we inline these variables to the cases in `when`, they are checked in sequence, producing three `is` checks). AFAIK, kotlin compiler doesn't try to do any optimizations for repeating logic in when cases.

Actually, the smart casts do work, but they cast to `TrieNode<*>` instead of `TrieNode<E>`. I also am not sure that having a smart cast avoids having a hard cast in backend code (I'm pretty sure in JVM it inserts a check anyways).

Reply:
Another option is to do `as?` instead of `is` and compare the result with null in when branches, e.g.
```
val thisAsNode = thisCell as TrieNode<E>
...
when {
    thisAsNode != null -> thisAsNode.doSomethingWithNode()
```